interfaces/apparmor/spec.go

```diff
@@ -334,21 +363,38 @@ func parent(path string) string {
 
 // Snippets returns a deep copy of all the added application snippets.
 func (spec *Specification) Snippets() map[string][]string {
-	return copySnippets(spec.snippets)
+	snippets := copySnippets(spec.snippets)
+	for tag, bag := range spec.dedupSnippets {
+		if bag != nil {
+			snippets[tag] = append(snippets[tag], bag.Items()...)
+		}
+	}
+	return snippets
```

Comment:
As I'm reading this, it seems like all the dedupSnippets will be at the end of the profile instead of, say, at the end of the rules for the interface, which is what I was expecting. If this is the only way to do it, we should comment on this above AddDeduplicatedSnippet() and either require an additional comment string as part of the snippet, or have that added programmatically.

Eg, as I'm reading this, we have:
```
... template rules ...

# Description: allow access to foo
... interface 'foo' rules

# Description: allow access to uio
... interface 'uio' rules for connection #1 

# Description: allow access to uio
... interface 'uio' rules for connection #2 

# Description: allow access to uio
... interface 'uio' rules for connection #N 

# Description: allow access to bar
... interface 'bar' rules

... dedup rules for uio
```

Ideally we would have:
```
... template rules ...

# Description: allow access to foo
... interface 'foo' rules

# Description: allow access to uio
... interface 'uio' rules for connection #1 

# Description: allow access to uio
... interface 'uio' rules for connection #2 

# Description: allow access to uio
... interface 'uio' rules for connection #N 

# Additional rules common for uio connections
... dedup rules for uio

# Description: allow access to bar
... interface 'bar' rules
```

but it would possibly be acceptable to have:
```
... template rules ...

# Description: allow access to foo
... interface 'foo' rules

# Description: allow access to uio
... interface 'uio' rules for connection #1 

# Description: allow access to uio
... interface 'uio' rules for connection #2 

# Description: allow access to uio
... interface 'uio' rules for connection #N 

# Description: allow access to bar
... interface 'bar' rules

# Additional rules common for uio connections
... dedup rules for uio
```

Reply:
If you look at the backend's `deriveContent` method and follow it to the specification's `SnippetForTag` you will see that we simply aggregate chunks that are otherwise devoid of any information as to which interface contributed them.

I think it would be fairly simple to change that so that we record tuples (<origin>, <snippet>) where origin would capture the plug or slot and any connection information.
We could then order the generated profiles such so that what you outlined above.

Reply:
If we aren't going to do this for 2.44, then we need to add a comment in dedupSnippets to know where it comes from. I'll comment in the right spot.